org/mozilla/jss/pkcs11/PK11Signature.c

```diff
@@ -45,87 +53,168 @@ JNIEXPORT void JNICALL
   (JNIEnv *env, jobject this)
 {
-    SGNContext *ctxt=NULL;
-    jobject contextProxy=NULL;
-    SECKEYPrivateKey *privk;
+    SGNContext *ctxt = NULL;
+    jobject contextProxy = NULL;
+    SECKEYPrivateKey *privk = NULL;
+    SECAlgorithmID *signAlg = NULL;
+    SECStatus rv = SECFailure;
+    PRArenaPool *arena = NULL;
+    SECOidTag signingAlg = SEC_OID_UNKNOWN;
 
     /* Extract the private key from the PK11Signature */
-    if( getPrivateKey(env, this, &privk) != PR_SUCCESS) {
-        PR_ASSERT( (*env)->ExceptionOccurred(env) != NULL);
+    if (getPrivateKey(env, this, &privk) != PR_SUCCESS) {
+        PR_ASSERT((*env)->ExceptionOccurred(env) != NULL);
         goto finish;
     }
 
-    /* Start the signing operation */
-    ctxt = SGN_NewContext(getAlgorithm(env, this), privk);
-	if(ctxt == NULL) {
-		JSS_throwMsg(env, TOKEN_EXCEPTION, "Unable to create signing context");
-		goto finish;
-	}
-    if( SGN_Begin(ctxt) != SECSuccess ) {
+    signingAlg = getAlgorithm(env,this);
+    if (signingAlg == SEC_OID_PKCS1_RSA_PSS_SIGNATURE) {
+        arena = PORT_NewArena(DER_DEFAULT_CHUNKSIZE);
+        if (!arena) {
+           JSS_throw(env, OUT_OF_MEMORY_ERROR);
+           goto finish;
+        }
+
+        rv = getRSAPSSParamsAndSigningAlg(env, this, arena, &signAlg, privk);
+        if (rv == SECFailure) {
+            goto finish;
+        }
+
+        /* Start the signing operation */
+        ctxt = SGN_NewContextWithAlgorithmID(signAlg, privk);
+    } else {
+        ctxt = SGN_NewContext(signingAlg, privk);
+    }
+
+    if (ctxt == NULL) {
+        if (arena != NULL) {
+            PORT_FreeArena(arena, PR_TRUE /* zero */);
+        }
+
+        JSS_throwMsg(env, TOKEN_EXCEPTION, "Unable to create signing context");
+        goto finish;
+    }
+
+    if (SGN_Begin(ctxt) != SECSuccess) {
         JSS_throwMsg(env, TOKEN_EXCEPTION, "Unable to begin signing context");
         goto finish;
     }
 
     /* Create a contextProxy and stick it in the PK11Signature object */
-    contextProxy = JSS_PK11_wrapSigContextProxy(env,
-												(void**)&ctxt,
-												SGN_CONTEXT);
-    if(contextProxy == NULL) {
-        PR_ASSERT( (*env)->ExceptionOccurred(env) != NULL);
+    contextProxy = JSS_PK11_wrapSigContextProxy(env, (void**)&ctxt,
+                                                SGN_CONTEXT, arena);
+
+    if (contextProxy == NULL) {
+        PR_ASSERT((*env)->ExceptionOccurred(env) != NULL);
         goto finish;
     }
+
+    // Signature algorithm for RSA PSS allocated in the arena,
+    // which is destroyed on exit.
     setSigContext(env, this, contextProxy);
 
 finish:
-    if(contextProxy==NULL && ctxt!=NULL) {
+    if (contextProxy == NULL && ctxt != NULL) {
         /* we created a context but not the Java wrapper, so we need to
          * delete the context here. */
         SGN_DestroyContext(ctxt, PR_TRUE /*freeit*/);
+        if (arena != NULL) {
+            PORT_FreeArena(arena, PR_TRUE /* zero */);
```

Comment:
When `ctx == NULL` and we're in an error case, we leak `arena`, meaning `signAlg` gets leaked.

Reply:
Yep, looks to be true.